himbaechel/uarch/gatemate/route_clock.cc

```diff
@@ -0,0 +1,162 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2024  The Project Peppercorn Authors.
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include <functional>
+#include <queue>
+
+#include "gatemate.h"
+#include "log.h"
+#include "nextpnr_assertions.h"
+
+#define HIMBAECHEL_CONSTIDS "uarch/gatemate/constids.inc"
+#include "himbaechel_constids.h"
+
+NEXTPNR_NAMESPACE_BEGIN
+
+namespace {
+
+struct QueuedWire
+{
+    explicit QueuedWire(WireId wire, float delay = 0.0) : wire{wire}, delay{delay} {};
+
+    WireId wire;
+    float delay;
+
+    bool operator>(const QueuedWire &rhs) const { return this->delay > rhs.delay; }
+};
+
+} // namespace
+
+void GateMateImpl::route_clock()
+{
+    auto clk_nets = std::vector<NetInfo *>{};
+
+    auto feeds_clk_port = [](PortRef &port) {
```

Comment:
do clocks for BRAMs etc feed through a CPE? or should they be considered here too?

Reply:
BRAMs have direct global CLK connection, so no need to handle it from here. Just CPEs need it, GPIO and BRAMs have those routed signals directly

Reply:
thanks for confirming!

Reply:
There are additional CLK inputs (going trough CPE) that we can use in case that is a problem.

Reply:
I think the only way we could feasibly discover that is with timing analysis...

Reply:
Indeed - just the first thing that comes to mind with an arrangement like this